### DISCUSSION TITLE
Help regarding plugins order and compinit

### DISCUSSION THREAD

**Discussion by xd003** [mattmc3/antidote#156] [Q&A] [answered]
Hello, i am pretty new to zsh altogether, please don't mind if this is too basic. I was reading the docs and noticed that you have emphasized on having a very optimal order of modules defined in ```.zsh_plugins.txt```. I just wanted to ensure that i have done it correctly as my system is bit on the lower end so i need to squeeze out as much performance/loading speed as possible. Below is my plugins txt file
```
romkatv/powerlevel10k
ohmyzsh/ohmyzsh path:plugins/gitfast
ohmyzsh/ohmyzsh path:plugins/colored-man-pages
zimfw/fzf
zdharma-continuum/fast-syntax-highlighting kind:defer
zsh-users/zsh-completions path:src kind:fpath
zsh-users/zsh-autosuggestions kind:defer
zsh-users/zsh-history-substring-search kind:defer
```
I just have 2 questions 
1) Would you recommend any change in the order of these  plugins or maybe the ```Kind``` annotations for these ?
2) I am not sure if any of the completion plugins load compinit for me. If not, where exactly should i add ```autoload -Uz compinit && compinit``` in my .zshrc ?

My current .zshrc is as follows
```
# Enable Powerlevel10k instant prompt. Should stay close to the top of ~/.config/zsh/.zshrc.
# Initialization code that may require console input (password prompts, [y/n]
# confirmations, etc.) must go above this block; everything else may go below.
if [[ -r "${XDG_CACHE_HOME:-$HOME/.cache}/p10k-instant-prompt-${(%):-%n}.zsh" ]]; then
  source "${XDG_CACHE_HOME:-$HOME/.cache}/p10k-instant-prompt-${(%):-%n}.zsh"
fi

# Set the name of the static .zsh plugins file antidote will generate.
zsh_plugins=${ZDOTDIR:-~}/.zsh_plugins.zsh

# Ensure you have a .zsh_plugins.txt file where you can add plugins.
[[ -f ${zsh_plugins:r}.txt ]] || touch ${zsh_plugins:r}.txt

# Lazy-load antidote.
fpath+=(${ZDOTDIR:-~}/.antidote)
autoload -Uz $fpath[-1]/antidote

# Generate static file in a subshell when .zsh_plugins.txt is updated.
if [[ ! $zsh_plugins -nt ${zsh_plugins:r}.txt ]]; then
  (antidote bundle <${zsh_plugins:r}.txt >|$zsh_plugins)
fi

# Source your static plugins file.
source $zsh_plugins

#### ALIASES #####
alias ls="eza -l --icons --hyperlink"

# To customize prompt, run `p10k configure` or edit ~/.config/zsh/.p10k.zsh.
[[ ! -f ~/.config/zsh/.p10k.zsh ]] || source ~/.config/zsh/.p10k.zsh
```

**Comment by mattmc3** [comment, marked as answer]
You've got it down for the most part. The only reason order matters with antidote is because many plugins are designed in such a way that order matters.

For example, from [zsh-history-substring-search](https://github.com/zsh-users/zsh-history-substring-search/tree/master#usage):
> If you want to use [zsh-syntax-highlighting](https://github.com/zsh-users/zsh-syntax-highlighting) along with this script, then make sure that you load it before you load this script

And [zsh-completions](https://github.com/zsh-users/zsh-completions) adds to your `fpath`, and Zsh requires your `fpath` to be fully built before running `compint`. There are a few `compint` plugins out there:
- mattmc3/zephyr path:plugins/completion
- belak/zsh-utils path:completion
- sorin-ionescu/prezto path:modules/completion

Of course, I'm biased towards [Zephyr](https://github.com/mattmc3/zephyr), but really if you include any of these towards the end of your `.zsh_plugins.txt`, your completions should work fine.

I'd also recommend removing `kind:defer` since you are already using P10k with insta-prompt. All the benefits of `zsh-defer` are to get plugins to load after your prompt is done, so it's great for Starship users, but P10k users already have this advantage.

```zsh
romkatv/powerlevel10k
ohmyzsh/ohmyzsh path:plugins/gitfast
ohmyzsh/ohmyzsh path:plugins/colored-man-pages
zimfw/fzf
zdharma-continuum/fast-syntax-highlighting

# completions
zsh-users/zsh-completions path:src kind:fpath
mattmc3/zephyr path:plugins/completion

# load last
zsh-users/zsh-autosuggestions
zsh-users/zsh-history-substring-search
```

One other note on completions. If you want to run compinit yourself and not rely on a plugin to do it, I recommend using [Zephyr's confd plugin](https://github.com/mattmc3/zephyr/blob/main/plugins/confd/confd.plugin.zsh) with antidote like so: `mattmc3/zephyr path:plugins/confd`.

`confd` behaves similar to how `Fish` loads configs, and runs any script it finds in a `${ZDOTDIR:-$HOME}/.zshrc.d` (or conf.d) folder. You could then add this script to that folder:

```zsh
# $ZDOTDIR/.zshrc.d/completions.zsh
autoload -Uz compinit && compinit
```

Your `.zsh_plugins.txt` then becomes:

```zsh
romkatv/powerlevel10k
ohmyzsh/ohmyzsh path:plugins/gitfast
ohmyzsh/ohmyzsh path:plugins/colored-man-pages
zimfw/fzf
zdharma-continuum/fast-syntax-highlighting

# completions
zsh-users/zsh-completions path:src kind:fpath

# conf.d - run compinit, and whatever else you want at this point in .zshrc.d
mattmc3/zephyr path:plugins/confd

# load last
zsh-users/zsh-autosuggestions
zsh-users/zsh-history-substring-search
```

Hope this helps.

**Reply by xd003** [reply]
Really comprehensive & detailed explanation, much appreciated, thanks. Just one last doubt, assuming i am using zephyr plugin for compinit , say If were to include another completion plugin https://github.com/marlonrichert/zsh-autocomplete for example, i should include it just above ```mattmc3/zephyr path:plugins/completion``` right ?

**Reply by mattmc3** [reply]
I haven't used that particular plugin much (though @marlonrichert makes some great stuff), but from those docs it looks like it may run compinit for you and wants you to load it before other plugins??

https://github.com/marlonrichert/zsh-autocomplete#manual-installation

I'm happy to try and help, but the specific behavior of plugins might be a better question for the plugin author, but let me know if it works because others that have the same question using antidote will hopefully find there way to this thread.

**Reply by xd003** [reply]
I was about to test that one out but didn't wanted to make it difficult to islolate my issue as i just realised maybe something is bit off.  It seems as if auto suggestions/completions are only working for a single session. Like say if i enter "echo test" and i try typing that again, i get the suggestion fine but if i close the terminal and type that again, i seem to get nothing 
```
romkatv/powerlevel10k
ohmyzsh/ohmyzsh path:plugins/gitfast
ohmyzsh/ohmyzsh path:plugins/colored-man-pages
zimfw/fzf
zdharma-continuum/fast-syntax-highlighting

# completions
zsh-users/zsh-completions path:src kind:fpath
mattmc3/zephyr path:plugins/completion

# load last
zsh-users/zsh-autosuggestions
zsh-users/zsh-history-substring-search
```
This is my current plugin list like you had recommended and my .zshrc is exactly same as the one posted initially in this discussion

**Reply by mattmc3** [reply]
Zsh has pretty terrible history defaults, and autosuggestions (and many plugins) rely on your history options and size to be configured properly. Have you tried using a history plugin like `mattmc3/zephyr path:plugins/history`?

**Reply by xd003** [reply]
Lemme test with this one, In what order should i use this one ideally ?
Edit - i also noticed something weird that zsh don't seem to be writing anything to the history file that i have defined in my .zshenv - ```export HISTFILE="$XDG_STATE_HOME/zsh/history"```, .zshenv is sourced correctly as echo $HISTFILE correctly returns the directory. Probably its because i am missing SAVEHIST & HISTSIZE vars, i will test with the plugin directly it once i know the order

**Reply by mattmc3** [reply]
You should probably change that to:

```zsh
export HISTFILE="${XDG_STATE_HOME:=$HOME/.local/state}/zsh/history"
[[ -e $HISTFILE:h ]] || mkdir -p $HISTFILE:h
```

Otherwise you run the risk of not making a proper history file.

**Reply by xd003** [reply]
i even tested with your history plugin but it seems that its unable to create the directory and history file for some reason, i am using zsh on Windows git bash at the moment. Given below is the location of the HISTFILE defined by the plugin
```
❯ HISTFILE="${XDG_DATA_HOME:-$HOME/.local/share}/zsh/zsh_history"
❯ cat $HISTFILE
cat: /c/Users/Admin/.local/share/zsh/zsh_history: No such file or directory
```

**Reply by mattmc3** [reply]
Have a look at my comment again:

> You should probably change that to:
> 
> ```shell
> export HISTFILE="${XDG_STATE_HOME:=$HOME/.local/state}/zsh/history"
> [[ -e $HISTFILE:h ]] || mkdir -p $HISTFILE:h
> ```
> 
> Otherwise you run the risk of not making a proper history file.

**Reply by xd003** [reply]
Thanks, that seemed to make it work. I guess it would be useful to add a note for new users just beside the completions section in FAQ
Something along the lines of - To make completions work for history, one needs to export ```HISTFILE```, ```SAVEHIST``` & ```HISTSIZE``` manually in their .zshenv/.zprofile or just simple use a plugin like ```mattmc3/zephyr path:plugins/history```
